### PR TITLE
Fix issue 19689: don't use recursive template parameter iteration in std.traits.

### PR DESCRIPTION
**Pull request by FeepingCreature**
Pull inner Impl templates out into std.traits so they're only instantiated once.
Use anySatisfy, staticMap instead of manually recursing.

Recursive template parameter iteration completely fails for large types. By using `staticMap`, we take advantage of its logarithmic recursion. `anySatisfy` is also inherently more efficient.

It's shorter too.

For an example of the problems the current approach causes, consider

```
import std.algorithm;
import std.format;
import std.range;

struct VeryLargeType
{
    static foreach (i; 500.iota)
    {
        mixin(format!"int v%s;"(i));
    }
}

void main()
{
    VeryLargeType a, b;
    move(a, b);
}
```

### CODE REVIEW

File: std/traits.d [hunk 1]
```diff
@@ -2902,42 +2874,55 @@ template RepresentationTypeTuple(T)
...
+    struct VeryLargeType
     {
-        static if (T.length == 0)
+        import std.format : format;
+        import std.range : iota;
+
+        static foreach (i; 500.iota)
         {
-            enum Impl = false;
+            mixin(format!"int v%s;"(i));
```
Comment (wilzbach) [inline std/traits.d]: How does this affect the compilation time of the unittests?

Comment (FeepingCreature) [reply std/traits.d]: Compilation time of just `std.traits`, averaged over 10 runs:

branch | time | change
---------|------|-----
master | 2.475s +- 28ms | 0%
PR with unittest | 2.589s +- 50ms | 5% slower
PR without unittest | 2.411s +- 22ms | 7% faster

Comment (wilzbach) [reply std/traits.d]: So it has a well-noticable effect.
Can you reproduce the error also with lower n?

Comment (FeepingCreature) [reply std/traits.d]: No, because it only happens when we hit DMD's template recursion limit, which is 500: see src/dmd/dsymbolsem.d:5817.

Comment (FeepingCreature) [reply std/traits.d]: ping

**Comment by FeepingCreature** [comment]
> Line 3613 or 3614

Ahaha yep, what the hell. Fixed. (Looks like a copypaste error from ~2012.)

**Pull request merged**